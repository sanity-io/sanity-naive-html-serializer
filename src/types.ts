import {ObjectField, SanityDocument, TypedObject, Schema} from 'sanity'

export type SerializedDocument = {
  name: string
  content: string
}

export type TranslationLevel = 'document' | 'field'
export interface Serializer {
  serializeDocument: (
    doc: SanityDocument,
    translationLevel: TranslationLevel,
    baseLang?: string,
    stopTypes?: string[],
    serializers?: Record<string, any>
  ) => SerializedDocument
  fieldFilter: (
    obj: Record<string, any>,
    objFields: ObjectField[],
    stopTypes: string[]
  ) => TypedObject
  languageObjectFieldFilter: (obj: Record<string, any>, baseLang: string) => Record<string, any>
  serializeArray: (
    fieldContent: Record<string, any>[],
    fieldName: string,
    stopTypes: string[],
    serializers: Record<string, any>
  ) => string
  serializeObject: (
    obj: TypedObject,
    stopTypes: string[],
    serializers: Record<string, any>
  ) => string
}

export type SerializerClosure = (schemas: Schema) => Serializer

export interface Deserializer {
  deserializeDocument: (
    serializedDoc: string,
    deserializers?: Record<string, any>,
    blockDeserializers?: Array<any>
  ) => Record<string, any>
  deserializeHTML: (
    html: string,
    deserializers: Record<string, any>,
    blockDeserializers: Array<any>
  ) => Record<string, any> | any[]
}

export interface Merger {
  fieldLevelMerge: (
    translatedFields: Record<string, any>,
    baseDoc: SanityDocument,
    localeId: string,
    baseLang: string
  ) => Record<string, any>
  documentLevelMerge: (
    translatedFields: Record<string, any>,
    baseDoc: SanityDocument
  ) => Record<string, any>
  reconcileArray: (origArray: any[], translatedArray: any[]) => any[]
  reconcileObject: (
    origObject: Record<string, any>,
    translatedObject: Record<string, any>
  ) => Record<string, any>
}
