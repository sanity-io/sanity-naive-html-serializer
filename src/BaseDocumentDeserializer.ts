import schema from 'part:@sanity/base/schema'
import Schema from '@sanity/schema'
import blockTools from '@sanity/block-tools'
import {
  customDeserializers,
  customBlockDeserializers,
} from './BaseSerializationConfig'
import { ObjectField, ObjectSchemaType, BlockSchemaType } from '@sanity/types'
import { Deserializer } from './types'

const noSchemaWarning = (obj: Element) =>
  `WARNING: Unfortunately the deserializer may have issues with this field or object: ${obj.className}.
  If it's a specific type, you may need to declare  at the top level, or write a custom deserializer.`

const defaultSchema = Schema.compile({
  name: 'default',
  types: [
    {
      type: 'object',
      name: 'default',
      fields: [
        {
          name: 'block',
          type: 'array',
          of: [{ type: 'block' }],
        },
      ],
    },
  ],
})

const blockContentType = defaultSchema
  .get('default')
  .fields.find((field: ObjectField) => field.name === 'block').type

//helper to handle messy input -- take advantage
//of blockTools' sanitizing behavior for single strings
const preprocess = (html: string) => {
  const intermediateBlocks = blockTools.htmlToBlocks(
    `<p>${html}</p>`,
    blockContentType
  )
  if (!intermediateBlocks.length) {
    throw new Error(`Error parsing string '${html}'`)
  }
  return intermediateBlocks[0].children[0].text
}

export const deserializeDocument = (
  serializedDoc: string,
  deserializers: Record<string, any> = customDeserializers,
  blockDeserializers = customBlockDeserializers
) => {
  const metadata: Record<string, any> = {}
  const head = new DOMParser().parseFromString(serializedDoc, 'text/html').head

  Array.from(head.children).forEach(metaTag => {
    const validTags = ['_id', '_rev', '_type']
    const metaName = metaTag.getAttribute('name')
    if (metaName && validTags.includes(metaName)) {
      metadata[metaName] = metaTag.getAttribute('content')
    }
  })

  const content: Record<string, any> = deserializeHTML(
    serializedDoc,
    schema.get(metadata._type),
    deserializers,
    blockDeserializers
  )

  return {
    ...metadata,
    ...content,
  }
}

const deserializeHTML = (
  html: string,
  target: ObjectSchemaType | BlockSchemaType,
  deserializers: Record<string, any>,
  blockDeserializers: Array<any>
): Record<string, any> | any[] => {
  //parent node is always div with classname of field -- get its children
  let HTMLnode = new DOMParser().parseFromString(html, 'text/html').body
    .children[0]
  if (!HTMLnode) {
    return {}
  }
  //@ts-ignore
  let output

  //catch embedded object as a field -- not necessarily foolproof
  if (
    HTMLnode.children.length === 1 &&
    target &&
    target.type &&
    HTMLnode.children[0].className === target.type.name
  ) {
    HTMLnode = HTMLnode.children[0]
  }

  //prioritize custom deserialization
  if (
    target &&
    target.type &&
    target.type.name &&
    Object.keys(deserializers.types).includes(target.type.name)
  ) {
    const deserialize = deserializers.types[target.type.name]
    output = deserialize(HTMLnode)
  }

  //begin recursively deserializing
  //every child node is either a field in an object
  //or a child object in an array
  else {
    const children = Array.from(HTMLnode.children)
    output = target && target.type && target.type.hasOwnProperty('of') ? [] : {}

    children.forEach(child => {
      let deserializedObj
      //prioritize custom deserialization
      if (Object.keys(deserializers.types).includes(child.className)) {
        const deserialize = deserializers.types[child.className]
        deserializedObj = deserialize(child)
      }

      //flat string, it's an unrich field
      else if (child.tagName.toLowerCase() === 'span') {
        deserializedObj = preprocess(child.innerHTML)
      }

      //has specific class name, so it's either a field or obj
      else if (child.className) {
        let objType
        //wrapper is fieldname
        if (target && target.fields) {
          objType = target.fields.find(field => field.name === child.className)
        }

        if (!objType && schema.get(child.className)) {
          objType = schema.get(child.className)
        }

        //@ts-ignore
        if (!objType && target && target.type && target.type.fields) {
          //@ts-ignore
          objType = target.type.fields.find(
            (field: ObjectField) => field.name === child.className
          )
        }

        if (!objType) {
          console.debug(noSchemaWarning(child))
          objType = blockContentType
        }

        try {
          deserializedObj = deserializeHTML(
            child.outerHTML,
            objType,
            deserializers,
            blockDeserializers
          )
        } catch (err) {
          console.debug(err)
          try {
            deserializedObj = deserializeHTML(
              child.outerHTML,
              schema.get('object'),
              deserializers,
              blockDeserializers
            )
          } catch (err) {
            console.debug(
              `Tried to deserialize block of type ${child.className} but failed! Received error ${err}`
            )
          }
        }

        // @ts-ignore
        if (Array.isArray(output) && deserializedObj) {
          deserializedObj._type = child.className
          deserializedObj._key = child.id
        }
      } else {
        deserializedObj = blockTools.htmlToBlocks(
          child.outerHTML,
          blockContentType,
          { rules: blockDeserializers }
        )[0]
        deserializedObj._key = child.id
      }

      // @ts-ignore
      if (Array.isArray(output)) {
        output.push(deserializedObj)
      } else {
        // @ts-ignore
        output[child.className] = deserializedObj
      }
    })
  }

  return output
}

export const BaseDocumentDeserializer: Deserializer = {
  deserializeDocument,
  deserializeHTML,
}
