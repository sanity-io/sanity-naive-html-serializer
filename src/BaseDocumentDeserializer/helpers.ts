import {htmlToBlocks} from '@sanity/block-tools'
import {Schema} from '@sanity/schema'
import {ObjectField, PortableTextSpan, PortableTextTextBlock} from 'sanity'

const defaultSchema = Schema.compile({
  name: 'default',
  types: [
    {
      type: 'object',
      name: 'default',
      fields: [
        {
          name: 'block',
          type: 'array',
          of: [{type: 'block'}],
        },
      ],
    },
  ],
})

export const blockContentType = defaultSchema
  .get('default')
  .fields.find((field: ObjectField) => field.name === 'block').type

export const noSchemaWarning = (obj: Element): string =>
  `WARNING: Unfortunately the deserializer may have issues with this field or object: ${obj.className}.
  If it's a specific type, you may need to declare  at the top level, or write a custom deserializer.`

//helper to handle messy input -- take advantage
//of blockTools' sanitizing behavior for single strings
export const preprocess = (html: string): string => {
  const intermediateBlocks = htmlToBlocks(
    `<p>${html}</p>`,
    blockContentType
  ) as PortableTextTextBlock<PortableTextSpan>[]
  if (!intermediateBlocks.length) {
    throw new Error(`Error parsing string '${html}'`)
  }
  return intermediateBlocks[0].children[0].text
}
