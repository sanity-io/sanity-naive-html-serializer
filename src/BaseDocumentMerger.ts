import {Merger} from './types'
import {SanityDocument} from 'sanity'
import {extractWithPath, arrayToJSONMatchPath} from '@sanity/mutator'

const reconcileArray = (origArray: any[], translatedArray: any[]): any[] => {
  //arrays of strings don't have keys, so just replace the array and return
  if (translatedArray && translatedArray.some((el) => typeof el === 'string')) {
    return translatedArray
  }

  //deep copy needed for field level patching
  const combined = JSON.parse(JSON.stringify(origArray))

  translatedArray.forEach((block) => {
    if (!block._key) {
      return
    }
    const foundBlockIdx = origArray.findIndex((origBlock) => origBlock._key === block._key)
    if (foundBlockIdx < 0) {
      //eslint-disable-next-line no-console
      console.warn(
        `This block no longer exists on the original document. Was it removed? ${JSON.stringify(
          block
        )}`
      )
    } else if (
      origArray[foundBlockIdx]._type === 'block' ||
      origArray[foundBlockIdx]._type === 'span'
    ) {
      combined[foundBlockIdx] = block
    } else if (Array.isArray(origArray[foundBlockIdx])) {
      combined[foundBlockIdx] = reconcileArray(origArray[foundBlockIdx], block)
    } else {
      //eslint-disable-next-line no-use-before-define -- this is a recursive function
      combined[foundBlockIdx] = reconcileObject(origArray[foundBlockIdx], block)
    }
  })
  return combined
}

const reconcileObject = (
  origObject: Record<string, any>,
  translatedObject: Record<string, any>
): Record<string, any> => {
  if (typeof translatedObject !== 'object' || !Object.keys(translatedObject).length) {
    return origObject
  }

  const updatedObj = JSON.parse(JSON.stringify(origObject))
  Object.entries(translatedObject).forEach(([key, value]) => {
    if (!value || key[0] === '_') {
      return
    }
    if (typeof value === 'string') {
      updatedObj[key] = value
    } else if (Array.isArray(value)) {
      //eslint-disable-next-line @typescript-eslint/no-use-before-define -- this is a recursive function
      updatedObj[key] = reconcileArray(origObject[key] ?? [], value)
    } else {
      updatedObj[key] = reconcileObject(origObject[key] ?? {}, value)
    }
  })
  return updatedObj
}

const fieldLevelMerge = (
  translatedFields: Record<string, any>,
  //should be fetched according to the revision and id of the translated obj above
  baseDoc: SanityDocument,
  localeId: string,
  baseLang: string = 'en'
): Record<string, any> => {
  const merged: Record<string, any> = {}
  const metaKeys = ['_rev', '_id', '_type']
  metaKeys.forEach((metaKey) => {
    if (translatedFields[metaKey]) {
      merged[metaKey] = translatedFields[metaKey]
    }
  })

  //get any field that matches the base language, because it's been translated
  const originPaths = extractWithPath(`..${baseLang}`, translatedFields)
  originPaths.forEach((match) => {
    const origVal = extractWithPath(arrayToJSONMatchPath(match.path), baseDoc)[0].value
    const translatedVal = extractWithPath(arrayToJSONMatchPath(match.path), translatedFields)[0]
      .value
    let valToPatch
    if (typeof translatedVal === 'string') {
      valToPatch = translatedVal
    } else if (Array.isArray(translatedVal) && translatedVal.length) {
      valToPatch = reconcileArray((origVal as Array<any>) ?? [], translatedVal)
    } else if (
      typeof translatedVal === 'object' &&
      Object.keys(translatedVal as Record<string, any>).length
    ) {
      valToPatch = reconcileObject(origVal ?? {}, translatedVal as Record<string, any>)
    }
    const destinationPath = [
      ...match.path.slice(0, match.path.length - 1), //cut off the "en"
      localeId.replace('-', '_'), // replace it with our locale
    ]

    merged[arrayToJSONMatchPath(destinationPath)] = valToPatch
  })

  return merged
}

const documentLevelMerge = (
  translatedFields: Record<string, any>,
  //should be fetched according to the revision and id of the translated obj above
  baseDoc: SanityDocument
): Record<string, any> => {
  return reconcileObject(baseDoc, translatedFields)
}

export const BaseDocumentMerger: Merger = {
  fieldLevelMerge,
  documentLevelMerge,
  reconcileArray,
  reconcileObject,
}
