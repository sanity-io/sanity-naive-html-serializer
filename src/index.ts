export {BaseDocumentMerger} from './BaseDocumentMerger'
export {BaseDocumentSerializer} from './BaseDocumentSerializer'
export {BaseDocumentDeserializer} from './BaseDocumentDeserializer'
export {
  defaultStopTypes,
  customSerializers,
  customBlockDeserializers,
} from './BaseSerializationConfig'

export type {SerializedDocument, Serializer, SerializerClosure, Deserializer, Merger} from './types'
