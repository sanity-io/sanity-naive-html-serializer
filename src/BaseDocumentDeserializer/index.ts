import BaseDocumentDeserializer from './BaseDocumentDeserializer'
import LegacyBaseDocumentDeserializer from './LegacyBaseDocumentDeserializer'

export { LegacyBaseDocumentDeserializer, BaseDocumentDeserializer }
