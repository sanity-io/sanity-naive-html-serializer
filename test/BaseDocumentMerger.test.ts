import { BaseDocumentMerger } from '../src/BaseDocumentMerger'
import clone from 'just-clone'
import { Block } from '@sanity/types'
import { getDeserialized } from './helpers'
const documentLevelArticle = require('./__fixtures__/documentLevelArticle')
const fieldLevelArticle = require('./__fixtures__/fieldLevelArticle')
const nestedLanguageFields = require('./__fixtures__/nestedLanguageFields')

let mockTestKey = 0

//needed to make snapshots happy on internal spans (where we don't track keys)
jest.mock('@sanity/block-tools/lib/util/randomKey.js', () => {
  return jest.fn().mockImplementation(() => {
    return `randomKey-${++mockTestKey}`
  })
})

beforeEach(() => {
  mockTestKey = 0
})

const getNewObject = () => {
  const newObject = {
    title: 'A new title',
    nestedArrayField: clone(documentLevelArticle.config.nestedArrayField),
    objectAsField: { title: 'A new nested title' },
    _key: null,
  }
  newObject.nestedArrayField[0].children[0].text = 'New text'
  return newObject
}

const getNewDocument = () => {
  const newDocument = getDeserialized(documentLevelArticle, 'document')
  newDocument.title = 'A new document title'
  newDocument.snippet = 'A new document snippet'
  newDocument.config = getNewObject()
  const newBlockText = newDocument.content[0]
  newBlockText.children[0].text = 'New block text'
  newDocument.content = [newBlockText]
  return newDocument
}

/*
 * DOCUMENT LEVEL
 */
describe('Document level merging', () => {
  const newDocument = getNewDocument()
  const mergedDocument = BaseDocumentMerger.documentLevelMerge(
    newDocument,
    documentLevelArticle
  )

  test('Global document level snapshot test', () => {
    expect(mergedDocument).toMatchSnapshot()
  })

  /*
   * Objects
   */
  test('Top-level string / text fields from new document override old document', () => {
    expect(mergedDocument.title).toEqual(newDocument.title)
    expect(mergedDocument.snippet).toEqual(newDocument.snippet)
    expect(mergedDocument.title).not.toEqual(documentLevelArticle.title)
    expect(mergedDocument.snippet).not.toEqual(documentLevelArticle.snippet)
  })

  test('Nested object fields override old object fields', () => {
    expect(mergedDocument.config.title).toEqual(newDocument.config.title)
    expect(mergedDocument.config.title).not.toEqual(
      documentLevelArticle.config.title
    )
    expect(mergedDocument.config.nestedArrayField[0].children[0].text).toEqual(
      newDocument.config.nestedArrayField[0].children[0].text
    )
    expect(
      mergedDocument.config.nestedArrayField[0].children[0].text
    ).not.toEqual(
      documentLevelArticle.config.nestedArrayField[0].children[0].text
    )
  })

  test('Nested object merge uses old fields when not present on new object', () => {
    expect(newDocument.config.objectAsField.content).toBeUndefined()
    expect(mergedDocument.config.objectAsField.content).toBeDefined()
  })

  /*
   * Arrays
   */
  test('Arrays will use new objects when they exist', () => {
    expect(mergedDocument.content[0].children[0].text).toEqual(
      newDocument.content[0].children[0].text
    )
    expect(mergedDocument.content[0].children[0].text).not.toEqual(
      documentLevelArticle.content[0].children[0].text
    )
  })

  test('Arrays will use old blocks if they do not exist on new object', () => {
    expect(newDocument.content[1]).toBeUndefined()
    expect(mergedDocument.content[1]).toBeDefined()
    expect(mergedDocument.content[1]._key).toEqual(
      documentLevelArticle.content[1]._key
    )
  })

  test('Arrays will merge objects in the array', () => {
    const newDocument = getNewDocument()
    const newObject = getNewObject()

    //add a new block with some new content, but not all new content
    newDocument.content.push({
      _key: documentLevelArticle.content[1]._key,
      objectAsField: newObject.objectAsField,
      title: newObject.title,
    })
    const mergedDocument = BaseDocumentMerger.documentLevelMerge(
      newDocument,
      documentLevelArticle
    )

    expect(mergedDocument.content[1].title).toEqual(
      newDocument.content[1].title
    )
    expect(mergedDocument.content[1].objectAsField.title).toEqual(
      newDocument.content[1].objectAsField.title
    )
    //content existed on old doc but not new, so the two coexist happily
    expect(newDocument.content[1].objectAsField.content).toBeUndefined()
    expect(mergedDocument.content[1].objectAsField.content).toBeDefined()
  })
})

/*
 * FIELD LEVEL -- todo (field level merge just wraps the other tested functions)
 */

const getNewFieldLevelObject = () => {
  const newObject = {
    title: 'A new title',
    nestedArrayField: clone(fieldLevelArticle.config.en.nestedArrayField),
    objectAsField: { title: 'A new nested title' },
    _key: null,
  }
  newObject.nestedArrayField[0].children[0].text = 'New text'
  return newObject
}

const getNewFieldLevelDocument = () => {
  const newDocument = getDeserialized(fieldLevelArticle, 'field')
  newDocument.title.en = 'A new document title'
  newDocument.snippet.en = 'A new document snippet'
  newDocument.config.en = getNewFieldLevelObject()
  const newBlockText = newDocument.content.en[0]
  newBlockText.children[0].text = 'New block text'
  newDocument.content.en = [newBlockText]
  return newDocument
}

describe('Field level merging', () => {
  const newDocument = getNewFieldLevelDocument()
  const fieldLevelPatches = BaseDocumentMerger.fieldLevelMerge(
    newDocument,
    fieldLevelArticle,
    'es_ES',
    'en'
  )
  test('Global field level snapshot test', () => {
    expect(fieldLevelPatches).toMatchSnapshot()
  })

  /*
   * Objects
   */
  test('Top-level string / text fields from new document patch to new field, and will be maintained in old field', () => {
    expect(fieldLevelPatches['title.es_ES']).toEqual(newDocument.title.en)
    expect(fieldLevelPatches['snippet.es_ES']).toEqual(newDocument.snippet.en)
    expect(fieldLevelPatches['title.es_ES']).not.toEqual(
      fieldLevelArticle.title.en
    )
    expect(fieldLevelPatches['snippet.es_ES']).not.toEqual(
      fieldLevelArticle.snippet.en
    )
    expect(fieldLevelPatches['title.en']).toBeUndefined()
    expect(fieldLevelPatches['snippet.en']).toBeUndefined()
  })

  test('Nested object fields override old object fields', () => {
    expect(fieldLevelPatches['config.es_ES'].title).toEqual(
      newDocument.config.en.title
    )
    expect(fieldLevelPatches['config.es_ES'].title).not.toEqual(
      fieldLevelArticle.config.en.title
    )
    expect(
      fieldLevelPatches['config.es_ES'].nestedArrayField[0].children[0].text
    ).toEqual(newDocument.config.en.nestedArrayField[0].children[0].text)
    expect(
      fieldLevelPatches['config.es_ES'].nestedArrayField[0].children[0].text
    ).not.toEqual(
      fieldLevelArticle.config.en.nestedArrayField[0].children[0].text
    )
  })

  test('Nested object merge uses old fields when not present on new object', () => {
    expect(fieldLevelPatches['config.es_ES'].objectAsField.content).toEqual(
      fieldLevelArticle.config.en.objectAsField.content
    )
  })

  /*
   * Arrays
   */
  test('Arrays will use new objects when they exist', () => {
    expect(fieldLevelPatches['content.es_ES'][0].children[0].text).toEqual(
      newDocument.content.en[0].children[0].text
    )
    expect(fieldLevelPatches['content.es_ES'][0].children[0].text).not.toEqual(
      documentLevelArticle.content[0].children[0].text
    )
  })

  test('Arrays will use old blocks if they do not exist on new object', () => {
    expect(newDocument.content.en[1]).toBeUndefined()
    expect(fieldLevelPatches['content.es_ES'][1]).toBeDefined()
    expect(fieldLevelPatches['content.es_ES'][1]._key).toEqual(
      fieldLevelArticle.content.en[1]._key
    )
  })

  test('Arrays will merge objects in the array', () => {
    const newDocument = getNewFieldLevelDocument()
    const newObject = getNewObject()

    //add a new block with some new content, but not all new content
    newDocument.content.en.push({
      _key: fieldLevelArticle.content.en[1]._key,
      objectAsField: newObject.objectAsField,
      title: newObject.title,
      //does not include "content" field -- we want that to be merged with the old
    })

    const fieldLevelPatches = BaseDocumentMerger.fieldLevelMerge(
      newDocument,
      fieldLevelArticle,
      'es_ES',
      'en'
    )

    expect(fieldLevelPatches['content.es_ES'][1].title).toEqual(
      newDocument.content.en[1].title
    )
    expect(fieldLevelPatches['content.es_ES'][1].objectAsField.title).toEqual(
      newDocument.content.en[1].objectAsField.title
    )
    //"content" field existed on old doc but not new, so the two coexist happily
    expect(newDocument.content.en[1].objectAsField.content).toBeUndefined()
    expect(
      fieldLevelPatches['content.es_ES'][1].objectAsField.content
    ).toBeDefined()
  })

  test('nested locale fields will be merged', () => {
    const newNestedFields = clone(nestedLanguageFields)
    newNestedFields.pageFields.name.en = 'This is a new page field name'
    newNestedFields.slices[0].en[0].children[0].text = 'This is new slice text'
    const baseDocument = { ...fieldLevelArticle, ...nestedLanguageFields }
    const newDocument = getDeserialized(
      { ...fieldLevelArticle, ...newNestedFields },
      'field'
    )
    const fieldLevelPatches = BaseDocumentMerger.fieldLevelMerge(
      newDocument,
      baseDocument,
      'es_ES',
      'en'
    )

    expect(fieldLevelPatches['slices[0].es_ES'][0].children[0].text).toEqual(
      newDocument.slices[0].en[0].children[0].text
    )
    expect(fieldLevelPatches['pageFields.name.es_ES']).toEqual(
      newDocument.pageFields.name.en
    )
  })
})

/*
 * STYLE TAGS
 */
test('Merged document should maintain style tags', () => {
  const newDocument = getNewDocument()
  const mergedDocument = BaseDocumentMerger.documentLevelMerge(
    newDocument,
    documentLevelArticle
  )
  const origH1Block = documentLevelArticle.content.find(
    (block: Block) => block.style === 'h1'
  )
  const origH2Block = documentLevelArticle.content.find(
    (block: Block) => block.style === 'h2'
  )
  const mergedH1Block = mergedDocument.content.find(
    (block: Block) => block.style === 'h1'
  )
  const mergedH2Block = mergedDocument.content.find(
    (block: Block) => block.style === 'h2'
  )
  expect(mergedH1Block).toBeDefined()
  expect(mergedH2Block).toBeDefined()
  expect(mergedH1Block._key).toEqual(origH1Block._key)
  expect(mergedH2Block._key).toEqual(origH2Block._key)
})
