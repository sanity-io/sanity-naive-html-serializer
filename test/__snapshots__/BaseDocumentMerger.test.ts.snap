// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document level merging Global document level snapshot test 1`] = `
Object {
  "_createdAt": "2021-09-01T23:00:13Z",
  "_id": "drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a",
  "_rev": "fxcnqj-uew-gr7-v6r-z965tfm69",
  "_type": "documentLevelArticle",
  "_updatedAt": "2021-09-01T23:02:05Z",
  "config": Object {
    "_type": "objectField",
    "nestedArrayField": Array [
      Object {
        "_key": "4a58adc7c507",
        "_type": "block",
        "children": Array [
          Object {
            "_key": "4f6bc15ae261",
            "_type": "span",
            "marks": Array [],
            "text": "New text",
          },
        ],
        "markDefs": Array [],
        "style": "normal",
      },
    ],
    "objectAsField": Object {
      "_type": "childObjectField",
      "content": Array [
        Object {
          "_key": "1aa228ac848b",
          "_type": "block",
          "children": Array [
            Object {
              "_key": "e2cf67af8e62",
              "_type": "span",
              "marks": Array [],
              "text": "This is a block text 2 levels deep",
            },
          ],
          "markDefs": Array [],
          "style": "normal",
        },
      ],
      "title": "A new nested title",
    },
    "title": "A new title",
  },
  "content": Array [
    Object {
      "_key": "a4d177e92666",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "randomKey-18",
          "_type": "span",
          "marks": Array [],
          "text": "New block text",
        },
      ],
      "markDefs": Array [],
      "style": "normal",
    },
    Object {
      "_key": "c0313627775e",
      "_type": "objectField",
      "objectAsField": Object {
        "_type": "childObjectField",
        "content": Array [
          Object {
            "_key": "4f24f6fbfae7",
            "_type": "block",
            "children": Array [
              Object {
                "_key": "d68f4288f5b7",
                "_type": "span",
                "marks": Array [],
                "text": "This is block text in a nested object in an object in top-level block text.",
              },
            ],
            "markDefs": Array [],
            "style": "normal",
          },
        ],
        "title": "This is a nested object in an object in top-level block text.",
      },
      "title": "This is an object in top-level block text.",
    },
    Object {
      "_key": "9e2ab13c6d63",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "17e856dc4766",
          "_type": "span",
          "marks": Array [],
          "text": "This is h1 text",
        },
      ],
      "markDefs": Array [],
      "style": "h1",
    },
    Object {
      "_key": "297142fbaf21",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "0b2f9e94a161",
          "_type": "span",
          "marks": Array [],
          "text": "This is h2 text",
        },
      ],
      "markDefs": Array [],
      "style": "h2",
    },
    Object {
      "_key": "76e648fc3845",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "49f466bbc78c",
          "_type": "span",
          "marks": Array [],
          "text": "Bullet 1",
        },
      ],
      "level": 1,
      "listItem": "bullet",
      "markDefs": Array [],
      "style": "normal",
    },
    Object {
      "_key": "d090cd8b27d2",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "599c6991018f",
          "_type": "span",
          "marks": Array [],
          "text": "nested bullet a",
        },
      ],
      "level": 2,
      "listItem": "bullet",
      "markDefs": Array [],
      "style": "normal",
    },
    Object {
      "_key": "1cdffa5d50f5",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "c16eb16f01cb",
          "_type": "span",
          "marks": Array [],
          "text": "Bullet 2",
        },
      ],
      "level": 1,
      "listItem": "bullet",
      "markDefs": Array [],
      "style": "h2",
    },
  ],
  "hidden": true,
  "meta": "Do not translate this",
  "snippet": "A new document snippet",
  "tags": Array [
    "tag 1",
    "tag 2",
    "tag 3",
  ],
  "title": "A new document title",
}
`;

exports[`Field level merging Global field level snapshot test 1`] = `
Object {
  "_id": "drafts.2947533e-1ea5-4116-955b-339608d3445d",
  "_rev": "l58oha-n06-1s4-f6i-74g6cuakl",
  "_type": "fieldLevelArticle",
  "config.es_ES": Object {
    "_type": "objectField",
    "nestedArrayField": Array [
      Object {
        "_key": "f49b4d7e3e51",
        "_type": "block",
        "children": Array [
          Object {
            "_key": "a6170f21181c",
            "_type": "span",
            "marks": Array [],
            "text": "New text",
          },
        ],
        "markDefs": Array [],
        "style": "normal",
      },
    ],
    "objectAsField": Object {
      "_type": "childObjectField",
      "content": Array [
        Object {
          "_key": "869168fe3a2a",
          "_type": "block",
          "children": Array [
            Object {
              "_key": "10d423df228a",
              "_type": "span",
              "marks": Array [],
              "text": "This is block text 2 levels deep",
            },
          ],
          "markDefs": Array [],
          "style": "normal",
        },
      ],
      "title": "A new nested title",
    },
    "title": "A new title",
  },
  "content.es_ES": Array [
    Object {
      "_key": "e2a39d768ff8",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "randomKey-62",
          "_type": "span",
          "marks": Array [],
          "text": "New block text",
        },
      ],
      "markDefs": Array [],
      "style": "normal",
    },
    Object {
      "_key": "271b0c6ee984",
      "_type": "objectField",
      "objectAsField": Object {
        "_type": "childObjectField",
        "content": Array [
          Object {
            "_key": "4f24f6fbfae7",
            "_type": "block",
            "children": Array [
              Object {
                "_key": "d68f4288f5b7",
                "_type": "span",
                "marks": Array [],
                "text": "This is block text in a nested object in an object in top-level block text.",
              },
            ],
            "markDefs": Array [],
            "style": "normal",
          },
        ],
        "title": "This is a nested object in an object in top-level block text.",
      },
      "title": "This is an object in top-level block text.",
    },
  ],
  "snippet.es_ES": "A new document snippet",
  "tags.es_ES": Array [
    "tag 1",
    "tag 2",
    "tag 3",
  ],
  "title.es_ES": "A new document title",
}
`;
